src/common/client.py

```
@@ -0,0 +1,79 @@
+            command="endpoint",
+            subcommand="status",
+            endpoints=self.client_url,
+            output_format="json",
+        ):
+            endpoint_status = json.loads(result)[0]
+
+        return endpoint_status
+
+    def _run_etcdctl(
+        self,
+        command: str,
+        subcommand: str | None,
+        endpoints: str,
+        output_format: str | None = "simple",
+    ) -> str | None:
+        """Execute `etcdctl` command via subprocess.
+
+        The list of arguments will be extended once authentication/encryption is implemented.
+        This method aims to provide a very clear interface for executing `etcdctl` and minimize
+        the margin of error on cluster operations.
+
+        Args:
+            command: command to execute with etcdctl, e.g. `elect`, `member` or `endpoint`
+            subcommand: subcommand to add to the previous command, e.g. `add` or `status`
+            endpoints: str-formatted list of endpoints to run the command against
+            output_format: set the output format (fields, json, protobuf, simple, table)
+            ...
+
+        Returns:
+            The output of the subprocess-command as a string. In case of error, this will
+            return `None`. It will not raise an error in order to leave error handling up
+            to the caller. Depending on what command is executed, the ways of handling errors
+            might differ.
+        """
+        try:
+            result = subprocess.run(
+                args=[
+                    "etcdctl",
+                    command,
+                    subcommand,
+                    f"--endpoints={endpoints}",
+                    f"-w={output_format}",
+                ],
+                check=True,
+                capture_output=True,
+                text=True,
+            ).stdout
+        except subprocess.CalledProcessError as e:
+            logger.warning(e)
+            return None
```

We should perhaps propagate the error here instead of returning `None`

---

I'm open for adjusting this, just want to give my reasoning here and would like to get your feedback on it: 

This function `_run_etcdctl` is very generic, as it will execute any command passed to it. The only error we will ever get back is the `CalledProcessError`, which would then have to be raised to the caller. We would then have to handle the same `CalledProcessError` in all places where we call `_run_etcdctl`. If we would only have to check the return, this would be much simpler: `if result := self._run_etcdctl(...).

Please let me know if you agree or not.

---

It's not an issue - however note that you will have to handle the result of your commands a bit more gracefully in the case of `result == ""` vs `result == None`

Meaning that you will not be able to perform checks like:
```
if not self._run_etcdctl(..):
   # this can mean Exception or success with no text result 
```

---

Yes I'm aware of this. That's why we have `capture_output=True` in the `subprocess` command, to always get something back.